Fix part of #5134: Increased Coverage from 82% to 100% for core.platform.email.gae_email_services.py

siddhantkhandelwal opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #5134: Increased Coverage from 82% to 100% for core.platform.email.gae_email_services.py

PTAL @DubeySandeep @seanlip 
Thanks
## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

siddhantkhandelwal committed:
Merge pull request #1 from oppia/develop

In Sync with oppia/develop

siddhantkhandelwal committed:
Fixes part of #5134: Increased Coverage from 82% to 100% for core.platform.email.gae_email_services.

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
             to=feconf.ADMIN_EMAIL_ADDRESS)
         self.assertEqual(1, len(messages))
+
+    def test_that_email_bcc_is_sent_when_bcc_admin_is_true(self):
```
If you're trying to ensure that an admin is bcc'd then you probably shouldn't be setting ADMIN_EMAIL_ADDRESS as the recipient. Otherwise, how do you know whether they received the email due to the bcc?

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        self.assertEqual(1, len(messages))
+
+    def test_that_email_reply_to_email_is_sent_when_reply_to_id_is_true(self):
```
I am not sure what this is testing. The assertion in the test does not seem to be checking anything to do with the reply-to id.

siddhantkhandelwal replied:
Recitified

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+            messages = self.mail_stub.get_sent_messages(
```
Pull this and the next statement out of the with clause. Ditto below.

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+        email_exception = (
+            self.assertRaisesRegexp(
+                ValueError,
+                'Malformed sender email address: %s' % malformed_sender_email)
```
Avoid as much logic in tests as possible; inline the sender email into the expected error string. Ditto below.

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+
+    def test_that_email_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (
```
Here and below, I suggest you just inline this, where it's possible to do so. Makes it easier to read, because the reader doesn't need to carry this variable around in their head.

siddhantkhandelwal replied:
I had missed on replying to this comment. Resolved. Thanks!

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+            messages_a = self.mail_stub.get_sent_messages(
```
Pull everything that does not need to be in the "with" outside it. Otherwise it becomes less clear which statement is actually causing the exception.

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+                to=self.RECIPIENT_B_EMAIL)
+            self.assertEqual(0, len(messages_b))
+
+    def test_that_bulk_mails_not_sent_if_recepient_email_is_malformed(self):
```
Think of other tests. For example, what if one recipient email is malformed and one isn't? Try and ensure that the tests specify fully the behaviour of the code (don't just think about coverage only).

siddhantkhandelwal replied:
Added other tests too. Thanks!

seanlip reviewed (changes requested):
Hi @siddhantkhandelwal, thanks for your PR. I took a first pass and left some comments, but please contact @DubeySandeep for subsequent reviews -- he'd be happy to help you with the starter projects. Thanks!

siddhantkhandelwal committed:
Fixes part of #5134: Increased Coverage from 82% to 100% for core.platform.email.gae_email_services.

seanlip commented:
Hi @siddhantkhandelwal, just a note: please double-check to ensure that every comment is addressed, and then request a review explicitly via a top-level comment (see instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change)). Thanks!

siddhantkhandelwal commented:
@seanlip I looked through your suggestions and made changes accordingly. Thanks for your review! I have put up a message on the Gitter Channel requesting for the review.

Thanks!

seanlip commented:
Thanks, @siddhantkhandelwal. @DubeySandeep will do the further review, and I'll take a final pass after that. To make the reviews faster please make sure that you have followed all the guidelines [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) beforehand.

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        valid_reply_to = gae_email_services.get_incoming_email_address(reply_to)
+        self.assertEqual(
+            messages[0].reply_to, valid_reply_to)
```
Can we move this to the line above (following max. 80 chars per line)?

siddhantkhandelwal replied:
Done.
Thanks!

siddhantkhandelwal reviewed (commented):
I believe I have addressed all the issues.
PTAL @DubeySandeep 
Thanks!

siddhantkhandelwal committed:
Merge pull request #2 from oppia/develop

Bring to Sync

siddhantkhandelwal committed:
Merge branch 'develop' into gae-email-services

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+        self.assertEqual(1, len(messages))
+        self.assertEqual(messages[0].bcc, feconf.ADMIN_EMAIL_ADDRESS)
+
+    def test_that_email_reply_to_email_is_sent_when_reply_to_id_is_true(self):
```
Do we have a boolean value for reply_to_id (..`_when_reply_to_id_is_true`)?

siddhantkhandelwal replied:
Oh, No. That was a silly mistake.
Corrected.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
+                'subject', 'body', 'html', bcc_admin=False,
```
remove bcc_admin if not required.

siddhantkhandelwal replied:
Removed
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        valid_reply_to = gae_email_services.get_incoming_email_address(reply_to)
+        self.assertEqual(
+            messages[0].reply_to, valid_reply_to)
```
Do we have any test where we are testing the reverse of this? i.e, without reply_to_id the message doesn't have any reply to id?

siddhantkhandelwal replied:
If the reply_to_id is not passed, the email_message.reply_to equals the value 'reply+reply_to_address@example.com'. I am not sure if it is supposed to behave like this. I can do assertEqual check for this string if you suggest so.

siddhantkhandelwal replied:
Similar to how bcc_admin is handled, it is not an attribute of the emai message object if bcc_admin is False, can we do something similar for reply_to too?

DubeySandeep replied:
> If the reply_to_id is not passed, the email_message.reply_to equals the value 'reply+reply_to_address@example.com'.

I don't think this is an expected behavior can you please go through it and see whether it's something we are doing or it's just from gae.
/cc @seanlip

seanlip replied:
That does seem a little odd. I think it's worth trying to figure out the source of this behaviour.

Note that the GAE code is open source (in fact, you have a copy locally in oppia_tools) so it should be possible to trace this.

siddhantkhandelwal replied:
@seanlip @DubeySandeep 
There was a mistake in the way I was testing this function. I had called the function `get_incoming_email_address` while asserting for reply_to_id.  I have corrected this. The gae module works as expected.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+        messages_b = self.mail_stub.get_sent_messages(to=self.RECIPIENT_B_EMAIL)
+        self.assertEqual(len(messages_b), 1)
+
+    def test_that_email_not_sent_if_can_send_emails_is_false(self):
```
Make it `bulk_email_not`...

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+        messages_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_A_EMAIL)
```
Can we move this to line above? (Max 80 char in a line)

siddhantkhandelwal replied:
I can do this, but another suggestion of yours prevents me from doing it.

> I think `to=self.RECIPIENT_EMAILS[0]` will make it more readable as the emails are sent to the `self.RECIPIENT_EMAILS`.

DubeySandeep replied:
Thanks for checking!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+        self.assertEqual(len(messages_a), 0)
+
+        messages_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_B_EMAIL)
```
Ditto.

siddhantkhandelwal replied:
I can do this, but another suggestion of yours prevents me from doing it.

> I think `to=self.RECIPIENT_EMAILS[0]` will make it more readable as the emails are sent to the `self.RECIPIENT_EMAILS`.

DubeySandeep replied:
Thanks for checking!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        messages_a = self.mail_stub.get_sent_messages(to=self.RECIPIENT_A_EMAIL)
```
I think `to=self.RECIPIENT_EMAILS[0]` will make it more readable as the emails are sent to the `self.RECIPIENT_EMAILS`.

siddhantkhandelwal replied:
Oh yes, makes much more sense.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
             to=feconf.ADMIN_EMAIL_ADDRESS)
         self.assertEqual(1, len(messages))
+
+    def test_that_email_bcc_is_sent_when_bcc_admin_is_true(self):
```
Do we have a test to ensure the reverse? (i.e, admin isn't bcc'd if bcc_admin=False)

siddhantkhandelwal replied:
Yes, I missed that.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,144 @@ def test_sending_email(self):
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
```
@seanlip, I'm not sure whether we should directly call this function in testing? 

![image](https://user-images.githubusercontent.com/16653571/53747916-4a289400-3eca-11e9-8a7c-e057607f07b6.png)
(From the unit-test perspective I think it's fine but wanted to have a confirmation.)

siddhantkhandelwal replied:
Oh yes, the comment does suggest against doing so. But it not being a private function, I assumed I can use it in writing Unit Test. If I am not supposed to do this, I believe we should change `send_bulk_mail` -> `_send_buk_mail` in the file gae_email_services.py to make this clear for other developers.

DubeySandeep replied:
> we should change send_bulk_mail -> _send_buk_mail in the file gae_email_services.py to make this clear for other developers.

`send_bulk_mail` cannot be made as a private function as it's used in the email_manager.py, this is more like a developer should go through the docstring before using a method.  

Also, I'm still not sure whether we should allow using this function for testing? (ditto for gae_email_services.sent_email) @seanlip, any suggestion?

seanlip replied:
It's fine to test it; it's part of the "public API" of this module.

siddhantkhandelwal replied:
Thanks @seanlip!

DubeySandeep reviewed (changes requested):
@siddhantkhandelwal, PR looks overall good to me have added a few comments (mostly style related), PTAL!

siddhantkhandelwal committed:
Added a couple of more tests. Minor styling changes.

siddhantkhandelwal reviewed (commented):
I have made some of the suggested changes. Others need to be discussed.
PTAL @DubeySandeep 
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,155 @@ def test_sending_email(self):
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+
+    def test_that_bulk_mails_not_sent_if_recepient_email_is_malformed(self):
```
make it 'recipient' (not `recepient`)

siddhantkhandelwal replied:
Done

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,155 @@ def test_sending_email(self):
+        self.assertEqual(len(message_b), 0)
+
+    def test_that_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+        malformed_sender_email = ''
```
Can we make this test follow this pattern:
- Try sending bulk email with the correct sender email
- check for expected behavior
- Try sending bulk email with invalid sender email
- ensure it raises an error!  

(ditto in other places, sorry I missed this in the last review.)

siddhantkhandelwal replied:
Bulk mail with correct sender_email is tested using:
`def test_that_correct_bulk_emails_sent(self):`

The test for false sender_email is separate. We can combine those two though. This will increase the length of these tests. I am not sure what I should do with this now.
Looking forward to your suggestions.
Thanks!

siddhantkhandelwal replied:
@DubeySandeep 
Any suggestions?
Thanks!

DubeySandeep replied:
@siddhantkhandelwal, through  `test_that_correct_bulk_emails_sent` you are testing the overall functionality, whereas through `test_that_bulk_mails_not_sent_if_sender_email_is_malformed` we are checking that an exception is raised under a certain condition, I was asking you to have a test to ensure that it doesn't raise any exception when email is correct. 

Also, I realized that all of the tests depends on correct email so I think we can leave it!

siddhantkhandelwal replied:
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,155 @@ def test_sending_email(self):
+        self.assertEqual(1, len(messages))
+        self.assertFalse(hasattr(messages, 'bcc'))
+
+    def test_that_email_has_reply_to_email_if_passed(self):
```
maybe we can make it `test _sending_email_with_reply_to_id_adds_reply_to_email`

siddhantkhandelwal replied:
Done

DubeySandeep reviewed (commented):
@siddhantkhandelwal, This mostly looks good to me, I have left a few comments PTAL!

siddhantkhandelwal committed:
Minor Changes

siddhantkhandelwal committed:
reply_to_id not passed test

DubeySandeep reviewed (commented):
@siddhantkhandelwal, LGTM!  

@seanlip, you can take a final look!

siddhantkhandelwal commented:
Thanks, @DubeySandeep for the review. Here is the screenshot of Coverage report as you had suggested:


![screenshot from 2019-03-08 00-10-36](https://user-images.githubusercontent.com/30776850/53982166-3be0af00-413a-11e9-8f4a-3512e3961704.png)

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
         self.assertEqual(1, len(messages))
+
+    def test_that_email_bcc_is_sent_if_bcc_admin_is_true(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
```
Before the action you should check that get_sent_messages() returns 0.

Ditto below. See the [guidelines for writing backend tests](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

siddhantkhandelwal replied:
Done!

DubeySandeep replied:
@siddhantkhandelwal, is this resolved?

DubeySandeep replied:
is this resolved? (if yes then please add a reply!)

siddhantkhandelwal replied:
I have put up a Baseline Verification Check here.
Done!

siddhantkhandelwal replied:
Yes, I have added the necessary Baseline verifications!

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
+                'subject', 'body', 'html', reply_to_id=reply_to)
```
reply_to --> reply_to_id

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+                'subject', 'body', 'html', reply_to_id=reply_to)
+        messages = self.mail_stub.get_sent_messages(
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        valid_reply_to = gae_email_services.get_incoming_email_address(reply_to)
```
Nope -- use a hardcoded string in the expectation below. Test interface, not implementation.

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+                'subject', 'body', 'html', bcc_admin=False)
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(1, len(messages))
+        self.assertFalse(hasattr(messages, 'bcc'))
```
This seems wrong, and ditto below. Why would messages (a list) have a 'bcc' attr?

siddhantkhandelwal replied:
Fixed

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+
+    def test_that_bulk_emails_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (self.assertRaisesRegexp(
```
Drop outer parens, they are not needed.

You might also consider just inlining this. It reduces the need to keep an additional variable in one's head while reading the code. We do similar things elsewhere so you can look at the existing code for examples. (Ditto below and elsewhere.)

siddhantkhandelwal replied:
Fixed

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+    def test_that_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+
+        malformed_sender_email = ''
+        email_exception = (self.assertRaisesRegexp(
```
Drop outer parens; consider inlining.

siddhantkhandelwal replied:
Fixed

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+        messages_b = self.mail_stub.get_sent_messages(
+            to=malformed_recipient_emails[1])
+        self.assertEqual(len(messages_b), 0)
+        malformed_recipient_emails = [self.RECIPIENT_A_EMAIL, '']
```
Add a newline above this, since it's basically a separate "chunk" of the test.

siddhantkhandelwal replied:
Done

seanlip reviewed (changes requested):
Thanks, @siddhantkhandelwal! I've taken a pass, and I think there are some errors.

Please work with @DubeySandeep until everything is resolved, and I will take a final look.  Thanks!

siddhantkhandelwal committed:
Testing method changes in some tests. Minor style changes

siddhantkhandelwal reviewed (commented):
I have made all the suggested changes.
PTAL @DubeySandeep @seanlip 
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+                feconf.SYSTEM_EMAIL_ADDRESS, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html', bcc_admin=True)
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
```
Can we move this to the line above?

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+        self.assertEqual(messages[0].bcc, feconf.ADMIN_EMAIL_ADDRESS)
+
+    def test_that_email_bcc_not_sent_if_bcc_admin_is_false(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
```
Follow this comment:

> Before the action you should check that get_sent_messages() returns 0.
> 
> Ditto below. See the guidelines for writing backend tests.

siddhantkhandelwal replied:
@DubeySandeep @seanlip 
Sorry for marking it as resolved, that was a mistake. Actually, I wanted to ask something about it.
I am not able to wrap my head around as to why I need to do that.
Though I have put up code for doing this in the latest commit.
It would be great if you could help me out here.
Thanks!

DubeySandeep replied:
@siddhantkhandelwal, We use to follow a pattern for writing backend tests (see [wiki page](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) for more info):

> 1. Tests should use the following general pattern:
> 
> (i) setup - this is where you prepare any inputs/environment needed for the test.
> (ii) baseline verification - check the values without performing any action.
> (iii) action - perform the action or function call that leads to the expected change.
> (iv) endline verification - check that the values in the baseline verification have changed accordingly.


having a (ii) baseline validation assures that the changes are made only via the (iii) action step.

DubeySandeep replied:
@siddhantkhandelwal, have you followed this comment, everywhere possible? (As asked by Sean)

I cannot see any such check inside `test_sending_email_with_reply_to_id_adds_reply_to_email` etc.

Please reply to the comment when you completely address them.

siddhantkhandelwal replied:
@DubeySandeep PTAL!
I have added Baseline Verification to the tests.
Thanks!

siddhantkhandelwal replied:
@dubeysandeep
PTAL

siddhantkhandelwal replied:
@DubeySandeep @seanlip 
PTAL!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        self.assertFalse(hasattr(messages[0], 'reply_to'))
+
+    def test_that_email_not_sent_if_sender_address_is_malformed(self):
```
Can we have a consistent test function name? (Remove `that` like above.)

siddhantkhandelwal replied:
I have removed `that` from every function definition. 
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
```
Can we move this to the line above? (Ditto everywhere else.)

siddhantkhandelwal replied:
I have fixed this wherever it was not violating the 80 characters Linting Rule.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
             gae_email_services.send_mail(
-                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
+                feconf.SYSTEM_EMAIL_ADDRESS,
+                feconf.ADMIN_EMAIL_ADDRESS,
```
Do we need this change?

siddhantkhandelwal replied:
I did that in order to maintain consistency in the script.
In other places, I had to split them to multiple lines so as to not violate the Lint Rules.
Will change it back if you say so.
Thanks!

DubeySandeep replied:
Is this "violate the Lint Rules"? If not then undo this change!

siddhantkhandelwal replied:
I have reverted the change.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+        # Tests that email has reply_to_id if passed.
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
```
Try using `self.RECIPIENT_EMAIL` instead of ADMIN for recipient.

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,170 @@ def test_sending_email(self):
+        self.assertEqual(messages[0].reply_to, expected_reply_to_address)
+
+    def test_sending_email_without_reply_to_id_not_add_reply_to_email(self):
+        # Tests that email does not have reply_to_id if not passed.
```
*reply_to email if reply_to_id is not passed.

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep reviewed (commented):
@siddhantkhandelwal, Thanks for the changes, I've left a few comments and I also released that few of the last review comments are not addressed. PTAL!

siddhantkhandelwal committed:
Style changes. Fixed missed test components. Changed Test function names

siddhantkhandelwal committed:
Style changes. Fixed missed test components. Changed Test function names

siddhantkhandelwal reviewed (commented):
I have made most of the changes.
Some of them need to be discussed.
PTAL @DubeySandeep @seanlip 
Thanks!

siddhantkhandelwal committed:
Merge pull request #3 from oppia/develop

Merge

siddhantkhandelwal committed:
Merge branch 'develop' into gae-email-services

siddhantkhandelwal committed:
Merge pull request #4 from oppia/develop

Update Fork

siddhantkhandelwal committed:
Merge branch 'develop' into gae-email-services

siddhantkhandelwal committed:
Added Baseline Assertion Checks, comments to some tests.

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -23,11 +23,20 @@
+
+    def setUp(self):
+        super(EmailTests, self).setUp()
+        self.signup(self.RECIPIENT_EMAIL, self.RECIPIENT_USERNAME)
```
is this required? (not sure can you please check!)

siddhantkhandelwal replied:
The tests run just fine without this explicit setUp method. I have removed them in the latest commit.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,201 @@ def test_sending_email(self):
+
+    def test_email_bcc_is_sent_if_bcc_admin_is_true(self):
+        """Tests that the email has ADMIN_EMAIL_ADDRESS as bcc
+        if bcc_admin is True.
```
Fit as much as word possible in the line above (following max 80 chars per line)

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,201 @@ def test_sending_email(self):
+        self.assertEqual(messages[0].reply_to, expected_reply_to_address)
+
+    def test_sending_email_without_reply_to_id_not_add_reply_to_email(self):
+        """Tests that email does not have reply_to_email
```
Ditto.

siddhantkhandelwal replied:
I have added the Baseline verifications for them!

DubeySandeep commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -36,10 +45,201 @@ def test_sending_email(self):
+        self.assertFalse(hasattr(messages[0], 'bcc'))
+
+    def test_sending_email_with_reply_to_id_adds_reply_to_email(self):
+        # Tests that email reply_to if passed.
```
Remove this line or move this to the place where you are passing the reply_to_id.

siddhantkhandelwal replied:
Done.
Thanks!

DubeySandeep reviewed (approved):
@siddhantkhandelwal, this looks good to me have left a few comments related to style PTAL!

Also, try adding reply to all the comments if you have addresed.

siddhantkhandelwal committed:
Minor changes, signup() removal in setUp for Classes

siddhantkhandelwal reviewed (commented):
@seanlip PTAL!
In the file gae_email_services.py, 
```
def send_mail(
        sender_email, recipient_email, subject, plaintext_body, html_body,
        bcc_admin=False, reply_to_id=None):
    """Sends an email. The client is responsible for recording any audit logs.

    In general this function should only be called from
    email_manager._send_email().

    Args:
        sender_email: str. The email address of the sender. This should be in
            the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
```
I am not sure as to what is the use of 'SENDER_NAME' in sender_email attribute. Is it supposed to store the SENDER_NAME, if yes, I am not sure as to how we deal it?

I tried passing the SENDER_NAME and <SENDER_EMAIL> in the sender_email attribute and the tests pass, however, the function invalid_email_reason as defined [here](https://cloud.google.com/appengine/docs/standard/python/refdocs/modules/google/appengine/api/mail#invalid_email_reason) does not validate it as such.

siddhantkhandelwal commented:
@seanlip
This needs your review. PTAL. 
Thanks!

DubeySandeep commented:
@siddhantkhandelwal, Thanks for pointing this, I had a discussion with Sean regarding the email format and we finalized to have our own email validation method which will check that the sender_email follows 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'. Can you please add this validation method and relevant test related to this new method? Thanks!

siddhantkhandelwal commented:
> @siddhantkhandelwal, Thanks for pointing this, I had a discussion with Sean regarding the email format and we finalized to have our own email validation method which will check that the sender_email follows 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'. Can you please add this validation method and relevant test related to this new method? Thanks!

That's great! I will work on this right away.
Thanks @DubeySandeep @seanlip !

siddhantkhandelwal committed:
Added the Sender Email Verification

siddhantkhandelwal commented:
@DubeySandeep @seanlip 

I have implemented the is_sender_email_valid function in gae_email_services.py and modified the test file gae_email_services_test.py appropriately.
However, Backend test are failing for:
- core.domain.skill_services_test: 1 errors, 0 failures
- core.domain.email_manager_test: 2 errors, 0 failures
- core.controllers.classifier_test: 0 errors, 1 failures

I have marked this PR as [WIP]. I will modify the appropriate files in order for these tests to pass.
It would be great if you could approve the current implementation of is_sender_email_valid in gae_email_services.py. PTAL.

Thanks!

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+        sender_email: str. The email address of the sender.
+
+    Returns:
+        False if sender_email is not valid.
```
`bool. Whether the sender_email is valid.`

siddhantkhandelwal replied:
Done Thanks!

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+    splitted_sender_email = sender_email.split(' ')
+    if len(splitted_sender_email) < 2:
+        return False
+
```
Do you need to check that the second part begins with "<" and ends with ">"?

siddhantkhandelwal replied:
Done! Thanks!

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
     return 'reply+%s@%s' % (reply_to_id, feconf.INCOMING_EMAILS_DOMAIN_NAME)
 
 
+def is_sender_email_valid(sender_email):
```
Prefix with underscore to mark it as module-private. For instructions on testing private functions, see point 9 [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

siddhantkhandelwal replied:
Done! Thanks!

seanlip reviewed (commented):
Hi @siddhantkhandelwal, I took a pass and it looks generally good. I left some comments. Thanks!

siddhantkhandelwal committed:
Made the function private. Switched the order of the tests.

DubeySandeep commented:
@siddhantkhandelwal, Note, backend tests are failing.

siddhantkhandelwal committed:
Merge pull request #5 from oppia/develop

Update Fork

DubeySandeep commented:
@siddhantkhandelwal, any update on this PR?

siddhantkhandelwal committed:
Merge branch 'develop' into gae-email-services

siddhantkhandelwal committed:
Fixed email_manager sender address

siddhantkhandelwal commented:
@DubeySandeep @seanlip PTAL.
Thanks!

siddhantkhandelwal commented:
@seanlip This needs your attention. PTAL.
Thanks!

DubeySandeep commented on core/domain/email_manager.py:
```diff
@@ -345,9 +345,10 @@ def send_job_failure_email(job_id):
     send_mail_to_admin(mail_subject, mail_body)
     other_recipients = (
         NOTIFICATION_EMAILS_FOR_FAILED_TASKS.value)
+    system_name_email = '%s <%s>' % ('SYSTEM', feconf.SYSTEM_EMAIL_ADDRESS)
```
Move this system name to feconf. Also, we are assuming the email to be in this pattern so maybe you can change `system_name_email` --> `system_email`.

DubeySandeep replied:
Moving system name to feconf will allow app owner to configure the system name.

seanlip replied:
I have a slightly different proposal here. Let's do:

`system_name_email = '%s <%s>' % (feconf.SYSTEM_EMAIL_NAME, SYSTEM_EMAIL_ADDRESS)

and let's have SYSTEM_EMAIL_NAME be ''. (We will update it to the right value on deployment.)

siddhantkhandelwal replied:
That's great. I will do the suggested changes. Thanks!

siddhantkhandelwal replied:
Done!

DubeySandeep commented on core/platform/email/gae_email_services.py:
```diff
@@ -63,13 +84,14 @@ def send_mail(
     if not feconf.CAN_SEND_EMAILS:
         raise Exception('This app cannot send emails.')
 
-    if not mail.is_email_valid(sender_email):
```
Can we also add this check into our own validator function? (Maybe in future if gae adds or changes their validator function?)

siddhantkhandelwal replied:
I will start working on this!
Thanks!

siddhantkhandelwal replied:
Done

DubeySandeep commented on core/platform/email/gae_email_services.py:
```diff
@@ -107,15 +129,15 @@ def send_bulk_mail(
     if not feconf.CAN_SEND_EMAILS:
         raise Exception('This app cannot send emails.')
 
-    if not mail.is_email_valid(sender_email):
```
Ditto.

siddhantkhandelwal replied:
Done

seanlip commented on core/domain/email_manager.py:
```diff
@@ -364,9 +365,9 @@ def send_mail_to_admin(email_subject, email_body):
     app_id = app_identity_services.get_application_id()
     body = '(Sent from %s)\n\n%s' % (app_id, email_body)
-
+    system_name_email = '%s <%s>' % ('SYSTEM', feconf.SYSTEM_EMAIL_ADDRESS)
```
Using literal constants is discouraged, especially if the usage is repeated. Move the constant to feconf as described above.

siddhantkhandelwal replied:
On it!

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+    Returns:
+        bool. Whether the sender_email is valid.
+    """
+    splitted_sender_email = sender_email.split(' ')
```
splitted --> split (or tokenized)

"splitted" isn't really a word.

siddhantkhandelwal replied:
Fixed

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+    if len(splitted_sender_email) < 2:
+        return False
+
+    email_address = splitted_sender_email[len(splitted_sender_email) - 1]
```
You can just do `[-1]`, right?

siddhantkhandelwal replied:
Definitely. :stuck_out_tongue:

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -63,13 +84,14 @@ def send_mail(
         raise ValueError(
             'Malformed recipient email address: %s' % recipient_email)
 
+    if not _is_sender_email_valid(sender_email):
```
I think, in previous comments (which you haven't replied to?), @DubeySandeep was talking about updating mail.is_email_valid in general, rather than doing something specific for sender emails as opposed to recipient emails. (@DubeySandeep, is that correct?)

DubeySandeep replied:
@seanlip, I had raised this issue with the docstring provided in the gae_email_services.
![image](https://user-images.githubusercontent.com/16653571/55704944-2988b800-59fb-11e9-9445-d5ea7aaa8f5e.png)

Although I had asked for any general email but in the above docstring, we had only asked for the sender email.
I'm not sure whether we should have such it for recipient email too? (As we only have the username (not the real name) of the users)

EDIT: I'm not sure whether we can have user's Name from their Google authentication API.

seanlip replied:
Oh, I see. Yes, good point -- thanks! I am happy with either just using the recipient email, or the username followed by the recipient email. I'm not sure which is better.

seanlip replied:
> EDIT: I'm not sure whether we can have user's Name from their Google authentication API.

No, let's not do that. See my previous response -- either is fine. Thanks for clarifying!

siddhantkhandelwal replied:
Thanks @seanlip @DubeySandeep for clarifying this.

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -23,23 +23,310 @@
+
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+        malformed_sender_email = 'malformed_sender_email@example.com'
```
Here and elsewhere, give a hint in the test data about **why** it's malformed. E.g. call this missing_email_name@example.com and the following one MissingEmailAddress. There are probably others you can clarify.

siddhantkhandelwal replied:
Sure. I will add comments explaining the reasons why these are malformed.
Thanks!

siddhantkhandelwal replied:
Done!

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -23,23 +23,310 @@
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+        malformed_sender_email = 'malformed_sender_email@example.com'
```
Ditto (re providing context for malformed-ness).

siddhantkhandelwal replied:
Done

seanlip reviewed (changes requested):
Hi @siddhantkhandelwal, this looks pretty good, thanks. Make sure to address all existing review comments explicitly. I've taken a full pass and there are only minor comments left.

siddhantkhandelwal committed:
Added custom email validator

siddhantkhandelwal commented:
Hey @seanlip @DubeySandeep 
I have made the suggested changes.
PTAL.

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+    Returns:
+        bool. Whether specified email address is valid.
+    """
+
```
Drop the unnecessary blank lines.

siddhantkhandelwal replied:
Done!

seanlip commented on core/platform/email/gae_email_services.py:
```diff
@@ -34,6 +34,27 @@ def get_incoming_email_address(reply_to_id):
+    """Determines whether an email address is invalid.
+
+    Args:
+        email_address: Email address to check.
```
Missing typeinfo in docstring.

siddhantkhandelwal replied:
Added.

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -120,9 +137,11 @@ def test_email_not_sent_if_sender_address_is_malformed(self):
         messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
         self.assertEqual(0, len(messages))
 
+        # Case when malformed_sender_email does not have name of the sender
```
Add comma after "sender".

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -146,9 +167,11 @@ def test_email_not_sent_if_sender_address_is_malformed(self):
         messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
         self.assertEqual(0, len(messages))
 
+        # Case when malformed_sender_email does not have email address of the
```
This comment doesn't make sense. Please proofread your comments and fix the wording.

siddhantkhandelwal replied:
Done

seanlip commented on core/platform/email/gae_email_services_test.py:
```diff
@@ -282,6 +350,8 @@ def test_bulk_mails_not_sent_if_sender_email_is_malformed(self):
             to=self.RECIPIENT_EMAILS[1])
         self.assertEqual(0, len(message_b))
 
+        # Case when malformed_sender_email does not have email address of the
```
Ditto.

siddhantkhandelwal replied:
Done

seanlip commented on core/domain/email_manager.py:
```diff
@@ -345,7 +345,8 @@ def send_job_failure_email(job_id):
         NOTIFICATION_EMAILS_FOR_FAILED_TASKS.value)
-    system_name_email = '%s <%s>' % ('SYSTEM', feconf.SYSTEM_EMAIL_ADDRESS)
+    system_name_email = '%s <%s>' % (
+        feconf.SYSTEM_NAME_EMAIL, feconf.SYSTEM_EMAIL_ADDRESS)
```
See my comment in the previous review: I wrote SYSTEM_EMAIL_NAME, not SYSTEM_NAME_EMAIL. Otherwise the variable looks like it represents an email.

siddhantkhandelwal replied:
Fixed

seanlip reviewed (changes requested):
Hi @siddhantkhandelwal, this PR is very close to done, but I need to ask you to pay closer attention to the review comments and to proofread your work so that we don't keep going round in reviews. PTAL at the inline comments and bear the above in mind for future PRs. Thanks!

siddhantkhandelwal committed:
Minor changes

seanlip reviewed (approved):
LGTM. Thanks!

siddhantkhandelwal commented:
@seanlip 
I have made the suggested changes. Sorry for missing out on the review comments. I will make sure this does not happen again. PTAL.
Thanks!

siddhantkhandelwal commented:
Thanks @seanlip for approving the changes!

siddhantkhandelwal commented:
@seanlip Can we please merge this now?
